certora/specs/CER-2-Operator/CER-68-Operator-authorization.spec

```diff
@@ -0,0 +1,142 @@
+    init_state axiom forall bytes19 prefix. ownerLookupGhost[prefix] == 0;
+}
+// makes sure the ownerLookupGhost is updated properly
+hook Sstore EthereumVaultConnectorHarness.ownerLookup[KEY bytes19 prefix] address value STORAGE {
+    ownerLookupGhost[prefix] = value;
+}
+// makes sure that reads from ownerLookup after havocs are correct
+hook Sload address value EthereumVaultConnectorHarness.ownerLookup[KEY bytes19 prefix] STORAGE {
+    require(ownerLookupGhost[prefix] == value);
+}
+
+// check that an owner of a prefix is always from that prefix
+invariant OwnerIsFromPrefix(bytes19 prefix)
+    getOwnerOf(prefix) == 0 || getAddressPrefix(getOwnerOf(prefix)) == prefix
+    filtered { 
+        f -> !isMustRevertFunction(f) &&
+        // We can't handle these functions since they have `CALL`s in them
+        f.selector != sig:batch(IEVC.BatchItem[] calldata).selector
+        && 
+        f.selector != sig:call(address, address, uint256, bytes calldata).selector
+    }
+
+/**
+ * Checks the inverse of the above rule: if an attacker tries to call
+ * `setOperator()`, the call reverts. We consider the caller an attacker if
+ * either if the prefix has no owner yet, but the caller is not from this prefix
+ * or if the prefix has an owner that is not the caller.
+ */
+rule theOwnerCanCallSetOperator() {
+    env e;
+
+    // this in an interesting way to revert... *shrug*
+    require(e.msg.value < nativeBalances[e.msg.sender]);
+
+    bytes19 addressPrefix;
+    address operator;
+    uint256 operatorBitField;
+
+    address owner = getOwnerOf(addressPrefix);
+    requireInvariant OwnerIsFromPrefix(addressPrefix);
+
+    // the actual caller (either msg.sender or the onBehalfOfAccount)
+    address caller = e.msg.sender;
+
+    if (e.msg.sender == currentContract) {
+        // we are within permit() and should use getOnBehalfOfAccount() instead
+        caller = getExecutionContextOnBehalfOfAccount(e);
+        // the owner is already set, not zero and not EVC
+        require(owner == caller && owner != 0 && owner != currentContract);
+        // the owner has the proper prefix
+        require(getAddressPrefix(owner) == addressPrefix);
+    } else {
+        // just a regular call from msg.sender
+        // msg.sender is from the proper prefix
+        require(getAddressPrefix(e.msg.sender) == addressPrefix);
+        // the owner is not set yet (zero) or identical to msg.sender
+        require(owner == 0 || owner == e.msg.sender);
+    }
+
+    // the operator can not be from the prefix either
+    require(getAddressPrefix(operator) != getAddressPrefix(caller));
+
+    // the current bitfield must be different from what we try to set
+    // [TODO] this is not part of the interface documentation yet
+    require(getOperator(addressPrefix, operator) != operatorBitField);
+
+    // call the setOperator() method.
+    setOperator@withrevert(e, addressPrefix, operator, operatorBitField);
+    satisfy(!lastReverted);
```

Comment:
so this rule checks that there is a case that doesn't revert. from the comment I was expect an assert . If we are just looking for a possible non reverting than we can remove many of the requires

Reply:
I updated this and made it an assert rule again.